lib/SPIRV/SPIRVLowerSaddWithOverflow.cpp

```diff
@@ -81,39 +77,85 @@ void SPIRVLowerSaddWithOverflowBase::visitIntrinsicInst(CallInst &I) {
-    return;
+  Function *ReplacementFunc = Mod->getFunction(FuncName);
+  if (!ReplacementFunc) { // This function needs linking.
+    Mod->getOrInsertFunction(FuncName, F.getFunctionType());
+    // Read LLVM IR with the intrinsic's implementation
+    SMDiagnostic Err;
+    auto MB = MemoryBuffer::getMemBuffer(LLVMSaddWithOverflow);
+    auto SaddWithOverflowModule =
+        parseIR(MB->getMemBufferRef(), Err, *Context,
+                [&](StringRef) { return Mod->getDataLayoutStr(); });
+    if (!SaddWithOverflowModule) {
+      std::string ErrMsg;
+      raw_string_ostream ErrStream(ErrMsg);
+      Err.print("", ErrStream);
+      SPIRVErrorLog EL;
+      EL.checkError(false, SPIRVEC_InvalidLlvmModule, ErrMsg);
+      return;
+    }
+
+    // Link in the intrinsic's implementation.
+    if (!Linker::linkModules(*Mod, std::move(SaddWithOverflowModule),
+                             Linker::LinkOnlyNeeded))
+      TheModuleIsModified = true;
+
+    ReplacementFunc = Mod->getFunction(FuncName);
+    assert(ReplacementFunc && "How did we not link in the necessary function?");
   }
-  FunctionCallee FC = Mod->getOrInsertFunction(FuncName, I.getFunctionType());
-  I.setCalledFunction(FC);
-
-  // Read LLVM IR with the intrinsic's implementation
-  SMDiagnostic Err;
-  auto MB = MemoryBuffer::getMemBuffer(LLVMSaddWithOverflow);
-  auto SaddWithOverflowModule =
-      parseIR(MB->getMemBufferRef(), Err, *Context,
-              [&](StringRef) { return Mod->getDataLayoutStr(); });
-  if (!SaddWithOverflowModule) {
-    std::string ErrMsg;
-    raw_string_ostream ErrStream(ErrMsg);
-    Err.print("", ErrStream);
-    SPIRVErrorLog EL;
-    EL.checkError(false, SPIRVEC_InvalidLlvmModule, ErrMsg);
-    return;
+
+  F.replaceAllUsesWith(ReplacementFunc);
+}
+
+void SPIRVLowerSaddWithOverflowBase::replaceSaddSat(Function &F) {
+  assert(F.getIntrinsicID() == Intrinsic::sadd_sat);
+
+  SmallVector<IntrinsicInst *, 4> Intrinsics;
+  for (User *U : F.users()) {
+    if (auto *II = dyn_cast<IntrinsicInst>(U))
+      Intrinsics.push_back(II);
   }
 
-  // Link in the intrinsic's implementation.
-  if (!Linker::linkModules(*Mod, std::move(SaddWithOverflowModule),
-                           Linker::LinkOnlyNeeded))
-    TheModuleIsModified = true;
+  // Get the corresponding sadd_with_overflow intrinsic for the sadd_sat.
+  Type *IntTy = F.getFunctionType()->getReturnType();
+  Function *SaddO =
+      Intrinsic::getDeclaration(Mod, Intrinsic::sadd_with_overflow, IntTy);
+
+  // Replace all uses of the intrinsic with equivalent code relying on
+  // sadd_with_overflow
+  IRBuilder<> Builder(F.getContext());
+  unsigned BitWidth = IntTy->getIntegerBitWidth();
+  Value *IntMin = Builder.getInt(APInt::getSignedMinValue(BitWidth));
+  Value *ShiftWidth = Builder.getIntN(BitWidth, BitWidth - 1);
+
+  for (IntrinsicInst *II : Intrinsics) {
+    Builder.SetInsertPoint(II);
+    // sadd_sat(a, b) => overflow ? (res >> bitwidth) ^ intmin : res;
```

Comment:
Can you please add a line to show where res and overflow are coming from? Thanks